qiskit/qasm3/grammar.py

```diff
@@ -0,0 +1,644 @@
+    """
+    indexIdentifier
+        : Identifier rangeDefinition
+        | Identifier ( LBRACKET expressionList RBRACKET )?
+        | indexIdentifier '||' indexIdentifier
+    """
+
+    pass
+
+
+class Expression(Class):
+    """
+    expression
+        // include terminator/unary as base cases to simplify parsing
+        : expressionTerminator
+        | unaryExpression
+        // expression hierarchy
+        | xOrExpression
+        | expression '|' xOrExpression
+    """
+
+    def __init__(self, something):
+        self.something = something  # TODO
+
+    def qasm(self):
+        return str(self.something)
+
+
+class IndexIdentifier2(IndexIdentifier):
+    """
+    indexIdentifier
+        : Identifier rangeDefinition
+        | Identifier ( LBRACKET expressionList RBRACKET )? <--
+        | indexIdentifier '||' indexIdentifier
+    """
+
+    def __init__(self, identifier: Identifier, expressionList: [Expression] = None):
+        self.identifier = identifier
+        self.expressionList = expressionList
+
+    def qasm(self):
+        if self.expressionList:
+            return f"{self.identifier.qasm()}[{', '.join([i.qasm() for i in self.expressionList])}]"
+        else:
+            return f"{self.identifier.qasm()}"
+
+
+class QuantumMeasurement(Class):
+    """
+    quantumMeasurement
+        : 'measure' indexIdentifierList
+    """
+
+    def __init__(self, indexIdentifierList: [Identifier]):
+        self.indexIdentifierList = indexIdentifierList
+
+    def qasm(self):
+        return f"measure {', '.join([i.qasm() for i in self.indexIdentifierList])};\n"
+
+
+class QuantumMeasurementAssignment(Statement):
+    """
+    quantumMeasurementAssignment
+        : quantumMeasurement ARROW indexIdentifierList
+        | indexIdentifier EQUALS quantumMeasurement  # eg: bits = measure qubits;
+    """
+
+    def __init__(self, indexIdentifier: IndexIdentifier2, quantumMeasurement: QuantumMeasurement):
+        self.indexIdentifier = indexIdentifier
+        self.quantumMeasurement = quantumMeasurement
+
+    def qasm(self):
+        return [f"{self.indexIdentifier.qasm()} = {self.quantumMeasurement.qasm()}"]
+
+
+class ExpressionTerminator(Expression):
+    """
+    expressionTerminator
+        : Constant
+        | Integer
+        | RealNumber
+        | booleanLiteral
+        | Identifier
+        | StringLiteral
+        | builtInCall
+        | kernelCall
+        | subroutineCall
+        | timingTerminator
+        | LPAREN expression RPAREN
+        | expressionTerminator LBRACKET expression RBRACKET
+        | expressionTerminator incrementor
+    """
+
+    pass
+
+
+class Integer(Expression):
+    """Integer : Digit+ ;"""
+
+    pass
+
+
+class Designator(Class):
+    """
+    designator
+        : LBRACKET expression RBRACKET
+    """
+
+    def __init__(self, expression: Expression):
+        self.expression = expression
+
+    def qasm(self):
+        return f"[{self.expression.qasm()}]"
```

Comment:
I'm not super familiar with the grammar, but should `Designator.qasm()` be responsible for generating the square brackets or `BitDeclaration` be responsible (iff it has a `designator`)?

Reply:
Done in 1ucian0#8